### PR TITLE
Fix #656, the maximum OBU size

### LINKED ISSUE #656: §3.2, §3.6: Max size of obu_header and audio_element_obu
|Version|Source|Clause(s)|
|:---:|:---:|:---:|
|AOM Working Group Draft, 17 July 2023|Apple|§3.2, §3.6|

#### Comment:
```
if (obu_extension_flag == 1) {
      leb128() extension_header_size;
      unsigned int (8*extension_header_size) extension_header_bytes;
    }
```
would allow 4GB extension header bytes. Any concerns here? Should it be restricted to a certain limit? Should it use fixed size length signaling which allows reasonable amount of size signaling?

Same question regarding audio_element_obu for param_definition_bytes and audio_element_config_bytes. which probably will not require max range of 4GB.

-----
#### Suggestion:
Clarify the range of OBU header and similar elements.

### PR DESCRIPTION
**Pull request by sunghee-hwang**
***
<a href="https://pr-preview.s3.amazonaws.com/AOMediaCodec/iamf/pull/724.html" title="Last updated on Aug 23, 2023, 5:57 AM UTC (89070fb)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/AOMediaCodec/iamf/724/7faeb6f...89070fb.html" title="Last updated on Aug 23, 2023, 5:57 AM UTC (89070fb)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -1724,7 +1724,8 @@ IA decoders SHALL be able to parse all OBUs explicitly listed for this version o
 
 NOTE: In this section and subsections, the meaning of a unique OBU is that it is still unique if it only varies by the [=obu_redundant_copy=] flag.
 
-Common restrictions on the [=IA Sequence=] for all profiles specified in this version of the specification:
+Common restrictions on the [=IA Sequence=] for all profiles specified in this version of the specification: \(\left( 1 + n \right)^2\), for \(n = 0, 1, 2, \ldots, 14\)
```
Comment (sunghee-hwang) [inline index.bs:1727]:
```suggestion
Common restrictions on the [=IA Sequence=] for all profiles specified in this version of the specification:
```

File: index.bs [hunk 2]
```diff
@@ -1725,6 +1725,7 @@ IA decoders SHALL be able to parse all OBUs explicitly listed for this version o
 NOTE: In this section and subsections, the meaning of a unique OBU is that it is still unique if it only varies by the [=obu_redundant_copy=] flag.
 
 Common restrictions on the [=IA Sequence=] for all profiles specified in this version of the specification:
+- The maximum size of an OBU (an obu_header followed by the OBU payload) SHALL be limited to \(2\text{MB}\) (i.e., \(2^21 \text{bytes}\)). It implies that the maximum value of the [=obu_size=] field SHALL be limited to \(2^21 - 4\).
```
Comment (sunghee-hwang) [inline index.bs:1728]:
```suggestion
- The maximum size of an OBU (an obu_header followed by the OBU payload) SHALL be limited to \(2\text{MB}\) (i.e., \(2^{21} \text{bytes}\)). It implies that the maximum value of the [=obu_size=] field SHALL be limited to \(2^{21} - 4\). 
```

File: index.bs [hunk 3]
```diff
@@ -1725,6 +1725,7 @@ IA decoders SHALL be able to parse all OBUs explicitly listed for this version o
 NOTE: In this section and subsections, the meaning of a unique OBU is that it is still unique if it only varies by the [=obu_redundant_copy=] flag.
 
 Common restrictions on the [=IA Sequence=] for all profiles specified in this version of the specification:
+- The maximum size of an OBU (an obu_header followed by the OBU payload) SHALL be limited to \(2\text{MB}\) (i.e., \(2^{21} \text{bytes}\)). It implies that the maximum value of the [=obu_size=] field SHALL be limited to \(2^{21} - 4\).
```
Comment (sunghee-hwang) [inline index.bs:1728]:
```suggestion
- The maximum size of an OBU (an obu_header followed by the OBU payload) SHALL be limited to \(2\text{MB}\) (i.e., \(2^{21}\) bytes). It implies that the maximum value of the [=obu_size=] field SHALL be limited to \(2^{21} - 4\). 
```

**Review by felicialim — APPROVED** [review]

File: index.bs [hunk 4]
```diff
@@ -1725,6 +1725,7 @@ IA decoders SHALL be able to parse all OBUs explicitly listed for this version o
 NOTE: In this section and subsections, the meaning of a unique OBU is that it is still unique if it only varies by the [=obu_redundant_copy=] flag.
 
 Common restrictions on the [=IA Sequence=] for all profiles specified in this version of the specification:
+- The maximum size of an OBU (an obu_header followed by the OBU payload) SHALL be limited to \(2\text{MB}\) (i.e., \(2^{21}\) bytes). It implies that the maximum value of the [=obu_size=] field SHALL be limited to \(2^{21} - 4\).
```
Comment (podborski) [inline index.bs:1728]: Will it not be 2^{21} - 5? Looks like 1 byte of header before the `obu_size` was not taken into consideration?

Comment (sunghee-hwang) [reply index.bs:1728]: 4 was based on "1 byte (obu header before obu_size) + 3 bytes for obu_size field".
Did I miss anything?

Comment (jwcullen) [reply index.bs:1728]: Also consider the the possibility that `obu_size` is encoded to be larger than the minimum size with leading zeroes. 

In an old version we referenced AV1's definition of `leb128` which explicitly allowed this. Although the current IAMF spec is less clear here.

Comment (podborski) [reply index.bs:1728]: @sunghee-hwang you are right, its 3 bytes I mistakenly thought you will require 4 for the size signaling. The point on the leading zeros is worth clarification too. I also made this comment https://github.com/AOMediaCodec/iamf/issues/687 as the `leb128` definition was not clear to me either.

Comment (sunghee-hwang) [reply index.bs:1728]: Let me create another PR to address #687.

**Pull request merged**